packages/flutter/lib/src/widgets/navigation_toolbar.dart

```diff
@@ -0,0 +1,133 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'package:flutter/rendering.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+
+/// [NavigationToolbar] is an unthemed layout helper to position 3 groups
+/// of widgets along a horizontal axis that's sensible for an application's
+/// navigation bar such as in Material Design and in iOS.
+///
+/// Optional [leading] and [trailing] widgets occupy the edges of the widget
+/// with reasonable constraints while the [middle] widget occupies the remaining
+/// space in either a center aligned or start aligned fashion.
+///
+/// Either directly use the themed app bars such as the Material [AppBar] or
+/// the iOS [CupertinoNavigationBar] or wrap this widget with more theming
+/// specifications for your own custom app bar.
+class NavigationToolbar extends StatelessWidget {
+  const NavigationToolbar({
+    Key key,
+    this.leading,
+    this.middle,
+    this.trailing,
+    this.centerMiddle: true,
+  }) : assert(centerMiddle != null),
+       super(key: key);
+
+  /// Widget to place at the start of the horizontal toolbar.
+  final Widget leading;
+  /// Widget to place in the middle of the horizontal toolbar, occupying
+  /// as much remaining space as possible.
+  final Widget middle;
+  /// Widget to place at the end of the horizontal toolbar.
+  final Widget trailing;
+  /// Whether to align the [middle] widget to the center of this widget or
+  /// next to the [leading] widget when false.
+  final bool centerMiddle;
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> children = <Widget>[];
+
+    if (leading != null)
+      children.add(new LayoutId(id: _ToolbarSlot.leading, child: leading));
+
+    if (middle != null)
+      children.add(new LayoutId(id: _ToolbarSlot.middle, child: middle));
+
+    if (trailing != null)
+      children.add(new LayoutId(id: _ToolbarSlot.trailing, child: trailing));
+
+    return new CustomMultiChildLayout(
+      delegate: new _ToolbarLayout(
+        centerMiddle: centerMiddle,
+      ),
+      children: children,
+    );
+  }
+}
+
+enum _ToolbarSlot {
+  leading,
+  middle,
+  trailing,
+}
+
+const double _kMiddleMargin = 16.0;
+
+class _ToolbarLayout extends MultiChildLayoutDelegate {
+  _ToolbarLayout({ this.centerMiddle });
```

Comment:
I might be wrong but it could be a bigger refactor, since the super has some non final fields.